### ISSUE TITLE
Can Tree Search run on GPU?

### ISSUE THREAD

**Issue by queliyong** [MolecularAI/aizynthfinder#40]
Hi,
When running several search at the same time, it costs much more time or even breaks down.  So I'm wondering if it can run on GPU to speed it up.
Thanks for any advice!

**Comment by SGenheden** [comment]
Hello

The expansion of molecules is done by a Keras/TF neural network, which of course can be computed on a GPU if you have setup your machine accordingly.
However, in my experience there is only a very small advantage of running these calculations on a GPU.

When you say that you are running several searches as the same time, can you be a little more specific? Are you using the `--nproc` argument to `aizynthcli`? With this options, I have noticed a bad performances on one specific machine on a supercluster. In that case, I manually split up my SMILES input and run many `aizynthcli` jobs in parallel without the `--nproc` argument. You can then concatenate the output with the `cat_aizynth_output` tool.

**Comment by queliyong** [comment]
> Hello
> 
> The expansion of molecules is done by a Keras/TF neural network, which of course can be computed on a GPU if you have setup your machine accordingly.
> However, in my experience there is only a very small advantage of running these calculations on a GPU.
> 
> When you say that you are running several searches as the same time, can you be a little more specific? Are you using the `--nproc` argument to `aizynthcli`? With this options, I have noticed a bad performances on one specific machine on a supercluster. In that case, I manually split up my SMILES input and run many `aizynthcli` jobs in parallel without the `--nproc` argument. You can then concatenate the output with the `cat_aizynth_output` tool.

'Running several searches as the same time' means that I open several aizynthapp GUI, and run search different target smiles

**Comment by queliyong** [comment]
> Hello
> 
> The expansion of molecules is done by a Keras/TF neural network, which of course can be computed on a GPU if you have setup your machine accordingly.
> However, in my experience there is only a very small advantage of running these calculations on a GPU.
> 
> When you say that you are running several searches as the same time, can you be a little more specific? Are you using the `--nproc` argument to `aizynthcli`? With this options, I have noticed a bad performances on one specific machine on a supercluster. In that case, I manually split up my SMILES input and run many `aizynthcli` jobs in parallel without the `--nproc` argument. You can then concatenate the output with the `cat_aizynth_output` tool.

'Running several searches as the same time' means that I open several aizynthapp GUI, and run search different target smiles individually at the same time. And it costs much more time to complete all of the searches.

**Comment by SGenheden** [comment]
I would recommend using `aizynthcli` if you are doing many predictions. But I don't know what you are trying to do or your setup. So that is just a general comment. 

I think your original question was answered above, so I'm inclined to close this ticket.

**Comment by queliyong** [comment]
> I would recommend using `aizynthcli` if you are doing many predictions. But I don't know what you are trying to do or your setup. So that is just a general comment.
> 
> I think your original question was answered above, so I'm inclined to close this ticket.

Thank you SG again for your help!
